File: fully_distributed/kMeans/src/main/java/it/unipi/hadoop/LocalConfiguration.java

package it.unipi.hadoop;

import java.io.File;
import java.io.IOException;
import org.apache.log4j.BasicConfigurator;
import org.ini4j.Wini;


public class LocalConfiguration {    
    // [Dataset]
    private long numberOfPoints;
    private int numberOfDimensions;
    private int numberOfClusters;
    private String inputPath;
    private String outputPath;
    
    // [K-means]
    private int seedRNG;
    private int clusteringNumberOfReduceTasks;
    private double distanceThreshold;
    private int maximumNumberOfIterations;
    
    // [Hadoop]
    private boolean verbose;
    
    public LocalConfiguration(String configPath) {
        BasicConfigurator.configure();
        
        try{
            Wini config = new Wini(new File(configPath));

            numberOfPoints = Long.parseLong(config.get("Dataset", "numberOfPoints"));
            numberOfDimensions = Integer.parseInt(config.get("Dataset", "numberOfDimensions"));
            numberOfClusters = Integer.parseInt(config.get("Dataset", "numberOfClusters"));
            inputPath = config.get("Dataset", "inputPath");
            outputPath = config.get("Dataset", "outputPath");
            seedRNG = Integer.parseInt(config.get("K-means", "seedRNG"));
            clusteringNumberOfReduceTasks = Integer.parseInt(config.get("K-means", "numberOfReduceTasks"));
            distanceThreshold = Double.parseDouble(config.get("K-means", "distanceThreshold"));
            maximumNumberOfIterations = Integer.parseInt(config.get("K-means", "maximumNumberOfIterations"));
            verbose = Boolean.valueOf(config.get("Hadoop", "verbose"));
            
            validateConfiguration();
        } catch(IOException e){
            System.err.println(e.getMessage());
            System.exit(1);
        }
    }
    
    private void validateConfiguration() {
        if (numberOfPoints <= 0) { 
            System.err.println("LocalConfiguration validation error: the number of points must be greater than 0");
            System.exit(1);
        }
        
        if (numberOfDimensions <= 0) { 
            System.err.println("LocalConfiguration validation error: the number of dimensions must be greater than 0");
            System.exit(1);
        }
        
        if (numberOfClusters <= 0) { 
            System.err.println("LocalConfiguration validation error: the number of clusters must be greater than 0");
            System.exit(1);
        }
        
        if (clusteringNumberOfReduceTasks <= 0) { 
            System.err.println("LocalConfiguration validation error: the number of reduce tasks must be greater than 0");
            System.exit(1);
        }
        
        if (distanceThreshold < 0) { 
            System.err.println("LocalConfiguration validation error: the error threshold must be greater than or equal to 0");
            System.exit(1);
        }
        
        if (maximumNumberOfIterations <= 0) { 
            System.err.println("LocalConfiguration validation error: the number of iterations must be greater than 0");
            System.exit(1);
        }
    }

    public long getNumberOfPoints() {
        return numberOfPoints;
    }

    public int getNumberOfDimensions() {
        return numberOfDimensions;
    }

    public int getNumberOfClusters() {
        return numberOfClusters;
    }

    public String getInputPath() {
        return inputPath;
    }

    public String getOutputPath() {
        return outputPath;
    }
    
    public int getSeedRNG() {
        return seedRNG;
    }
    
    public int getClusteringNumberOfReduceTasks() {
        return clusteringNumberOfReduceTasks;
    }
    
    public double getDistanceThreshold() {
        return distanceThreshold;
    }
    
    public int getMaximumNumberOfIterations() {
        return maximumNumberOfIterations;
    }
    
    public boolean getVerbose() {
        return verbose;
    }
     
    public void printConfiguration() {
        System.out.println("numberOfPoints = " + numberOfPoints);
        System.out.println("numberOfDimensions = " + numberOfDimensions);
        System.out.println("numberOfClusters = " + numberOfClusters);
        System.out.println("inputPath = " + inputPath);
        System.out.println("outputPath = " + outputPath);
        System.out.println("seedRNG = " + seedRNG);
        System.out.println("clusteringNumberOfReduceTasks = " + clusteringNumberOfReduceTasks);
        System.out.println("distanceThreshold = " + distanceThreshold);
        System.out.println("maximumNumberOfIterations = " + maximumNumberOfIterations);
        System.out.println("verbose = " + String.valueOf(verbose));
        System.out.println("");
    }
}

File: kMeans/LocalConfiguration.py
import configparser as cp


class LocalConfiguration:

    def __init__(self, configuration_path):
        self.configuration = cp.ConfigParser()
        self.configuration.read(configuration_path)

        # Dataset
        self.input_path = self.configuration.get("Dataset", "inputPath")
        self.output_path = self.configuration.get("Dataset", "outputPath")
        self.number_of_clusters = self.configuration["Dataset"].getint("numberOfClusters")
        self.number_of_points = self.configuration["Dataset"].getint("numberOfPoints")

        # K-means
        self.seed_RNG = self.configuration["K-means"].getint("seedRNG")
        self.distance_threshold = self.configuration["K-means"].getfloat("distanceThreshold")
        self.max_number_iterations = self.configuration["K-means"].getint("maximumNumberOfIterations")

        # Spark
        self.log_level = self.configuration.get("Spark", "logLevel")

        self.__validate()

    def __validate(self):
        if self.number_of_points <= 0:
            print("LocalConfiguration validation error: the number of points must be greater than 0")
            exit()

        if self.number_of_clusters <= 0:
            print("LocalConfiguration validation error: the number of clusters must be greater than 0")
            exit()

        if self.distance_threshold < 0:
            print("LocalConfiguration validation error: the distance threshold must be greater than or equal to 0")
            exit()

        if self.max_number_iterations <= 0:
            print("LocalConfiguration validation error: the number of iterations must be greater than 0")
            exit()

    def get_input_path(self):
        return self.input_path

    def get_output_path(self):
        return self.output_path

    def get_number_of_points(self):
        return self.number_of_points

    def get_number_of_clusters(self):
        return self.number_of_clusters

    def get_seed_RNG(self):
        return self.seed_RNG

    def get_log_level(self):
        return self.log_level

    def get_distance_threshold(self):
        return self.distance_threshold

    def get_maximum_number_of_iterations(self):
        return self.max_number_iterations

    def print(self):
        print("numberOfPoints = " + str(self.number_of_points))
        print("numberOfClusters = " + str(self.number_of_clusters))
        print("inputPath = " + self.input_path)
        print("outputPath = " + self.output_path)
        print("seedRNG = " + str(self.seed_RNG))
        print("distanceThreshold = " + str(self.distance_threshold) + "%")
        print("maximumNumberOfIterations = " + str(self.max_number_iterations))
        print("logLevel = " + self.log_level)